pkg/steps/git_source.go

```diff
@@ -0,0 +1,140 @@
+}
+
+func (s *gitSourceStep) Requires() []api.StepLink { return nil }
+
+func (s *gitSourceStep) Creates() []api.StepLink {
+	return []api.StepLink{api.InternalImageLink(s.gitConfig.To)}
+}
+
+func (s *gitSourceStep) Provides() (api.ParameterMap, api.StepLink) {
+	return api.ParameterMap{
+		"BUILD_ROOT_IMAGE": func() (string, error) {
+			is, err := s.imageClient.ImageStreams(s.jobSpec.Namespace()).Get(PipelineImageStream, meta.GetOptions{})
+			if err != nil {
+				return "", fmt.Errorf("could not get output imagestream: %v", err)
+			}
+			var registry string
+			if len(is.Status.PublicDockerImageRepository) > 0 {
+				registry = is.Status.PublicDockerImageRepository
+			} else if len(is.Status.DockerImageRepository) > 0 {
+				registry = is.Status.DockerImageRepository
+			} else {
+				return "", fmt.Errorf("image stream %s has no accessible image registry value", s.gitConfig.To)
+			}
+			return fmt.Sprintf("%s:%s", registry, s.gitConfig.To), nil
+		},
+	}, api.InternalImageLink("root")
+}
+
+func buildFromGitSource(jobSpec *JobSpec, gitConfig api.InputGitSourceStepConfiguration, toTag api.PipelineImageStreamTagReference, resources api.ResourceConfiguration) *buildapi.Build {
+	log.Printf("Building %s", toTag)
+	buildResources, err := resourcesFor(resources.RequirementsForStep(string(toTag)))
+	if err != nil {
+		panic(fmt.Errorf("unable to parse resource requirement for build %s: %v", toTag, err))
+	}
+
+	layer := buildapi.ImageOptimizationSkipLayers
+	build := &buildapi.Build{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      string(toTag),
+			Namespace: jobSpec.Namespace(),
+			Labels: map[string]string{
+				PersistsLabel:    "false",
+				JobLabel:         jobSpec.Job,
+				BuildIdLabel:     jobSpec.BuildId,
+				ProwJobIdLabel:   jobSpec.ProwJobID,
+				CreatesLabel:     string(toTag),
+				CreatedByCILabel: "true",
+			},
+			Annotations: map[string]string{
+				JobSpecAnnotation: jobSpec.rawSpec,
+			},
+		},
+		Spec: buildapi.BuildSpec{
+			CommonSpec: buildapi.CommonSpec{
+				Resources:      buildResources,
+				ServiceAccount: "builder",
+				Source: buildapi.BuildSource{
+					Type:       buildapi.BuildSourceGit,
+					ContextDir: gitConfig.BaseImage.ContextDir,
+					Git: &buildapi.GitBuildSource{
+						URI: gitConfig.BaseImage.Uri,
+						Ref: gitConfig.BaseImage.Ref,
+					},
+				},
+				Strategy: buildapi.BuildStrategy{
+					Type: buildapi.DockerBuildStrategyType,
+					DockerStrategy: &buildapi.DockerBuildStrategy{
+						DockerfilePath: gitConfig.BaseImage.Dockerfile,
+						ForcePull:      true,
+						NoCache:        true,
+						Env:            []coreapi.EnvVar{},
+						ImageOptimizationPolicy: &layer,
+					},
+				},
+				Output: buildapi.BuildOutput{
+					To: &coreapi.ObjectReference{
+						Kind:      "ImageStreamTag",
+						Namespace: jobSpec.Namespace(),
+						Name:      fmt.Sprintf("%s:%s", PipelineImageStream, toTag),
+					},
+				},
+			},
+		},
+	}
```

Comment:
Only the `BuildSource` is different from the existing source build, copying the code creates a lot of unnecessary duplication.

Reply:
Also, there is a slight difference in the strategy of the build. I split them up because I didn't want to construct 2 different things with the same method, and add checks and change the source and strategy depending on which in nil or not.